apm-agent-core/src/main/java/co/elastic/apm/agent/impl/metadata/Agent.java

```diff
@@ -75,4 +85,56 @@ public String getVersion() {
     public String getEphemeralId() {
         return ephemeralId;
     }
+
+    public String getActivationMethod() {
```

Comment:
Please do the discovery during instance construction unless there is a good reason to do lazily. 
All metadata discovery is happening on construction and everything that needs metadata is blocking until it is ready. This introduces concurrency in discovery that we don't have otherwise.

Reply:
Full metadata discovery doesn't happen at instance construction. The agent is part of the service object which does discovery in a queue of various discoveries. At the time of the Agent object creation, the tracer is not yet initialized, but that is needed for the activation discovery. The lazy initialization is needed, and guaranteed to work as the metadata is first called when it is serialized which explicitly happens after discovery is complete